add normalizer, tokenizer to rouge

hookSSi opened the pull request:
## What does this PR do?

Fixes #825 
I wrote some code but I don't know how to test it sorry.
plz review me hard

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

Borda commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -101,17 +106,31 @@ def _normalize_and_tokenize_text(text: str, stemmer: Optional[Any] = None) -> Se
+        # Replace any non-alpha-numeric characters with spaces.
+        text = re.sub(r"[^a-z0-9]+", " ", text.lower())
+
+    if tokenizer:
+        tokens = tokenizer.tokenize(text)
+    else:
+        tokens = re.split(r"\s+", text)
```
```suggestion
    tokens = tokenizer.tokenize(text) if tokenizer else re.split(r"\s+", text)
```

Borda reviewed (commented):
@stancld mind help with adding relevant tests? :)

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -101,17 +106,31 @@ def _normalize_and_tokenize_text(text: str, stemmer: Optional[Any] = None) -> Se
     """
-    # Replace any non-alpha-numeric characters with spaces.
-    text = re.sub(r"[^a-z0-9]+", " ", text.lower())
+    if normalizer:
+        text = normalizer.normalize(text)
+    else:
+        # Replace any non-alpha-numeric characters with spaces.
+        text = re.sub(r"[^a-z0-9]+", " ", text.lower())
+
+    if tokenizer:
+        tokens = tokenizer.tokenize(text)
+    else:
+        tokens = re.split(r"\s+", text)
```
Wondering if a requirement that each normalizer/tokenizer have `normalize`/`tokenize` method respectively is not too strict. Wouldn't be better to have something like
```suggestion
    if normalizer:
        text = normalizer(text)
    else:
        # Replace any non-alpha-numeric characters with spaces.
        text = re.sub(r"[^a-z0-9]+", " ", text.lower())

    if tokenizer:
        tokens = tokenizer(text)
    else:
        tokens = re.split(r"\s+", text)
```
i.e. normalizer/tokenizer is callable? Also, with this implementation, any method can be easily wrapped with a `partial` by a user?
Any thoughts @Borda?

Borda replied:
that looks nice, but don't have a strong pref...
@SkafteNicki ?

hookSSi replied:
I thought so at first, but I thought it was better to receive and process it as an object in optimizing even if it was a little more strict than to create a regular expression object or something in every function.

stancld replied:
In general, I'm open to both options. Maybe, can you add an example using own tokenizer and normalizer into `tm_examples` as it is the case, for example, for `BERTScore`? I think it can be beneficial for this PR as the usage of own tokenizer and normalizer is not included in the reference `rouge_score` package :]

hookSSi replied:
umm... I will try then!

Borda commented:
@stancld how is it going here? :)

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -101,17 +106,27 @@ def _normalize_and_tokenize_text(text: str, stemmer: Optional[Any] = None) -> Se
             An input sentence.
         stemmer:
             Porter stemmer instance to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer instance.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This instance must have method named ``normalize``. This method must take a string and return a string.
+        tokenizer:
+            A user's own tokenizer instance. If this is none, `spliting by spaces` is default
+            This instance must have method named ``tokenize``. This method must take a string and return `List[str]`
```
According to the example and current implementation, there are no more `nomalize` and `tokenize` methods. Also, I would mention it might be just a plain function.

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -306,6 +334,13 @@ def rouge_score(
             - ``best`` takes the best fmeasure score obtained between prediction and multiple corresponding references.
         use_stemmer:
             Use Porter stemmer to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer instance.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This instance must have method named ``normalize``. This method must take a string and return a string.
+        tokenizer:
+            A user's own tokenizer instance. If this is none, `spliting by spaces` is default
+            This instance must have method named ``tokenize``. This method must take a string and return `List[str]`
```
^^ same update here

hookSSi replied:
@stancld thank you for your update!

stancld reviewed (commented):
@hookSSi Thanks a lot for your contribution. I left a few minor comments but, in general, I think the PR looks quite good.
Regarding the tests.. I think we can just test default behaviour against the scenario we input the default tokenizer and normalizer manually into the module/functional metric and just verify everything goes through without problems. :]

Borda commented:
@hookSSi how is it going here? have you had a change to check recent review/suggestion from @stancld :rabbit:

hookSSi commented:
> @hookSSi how is it going here? have you had a change to check recent review/suggestion from @stancld 🐰

Sorry, I'm checking on it now

stancld commented:
> > @hookSSi how is it going here? have you had a change to check recent review/suggestion from @stancld 🐰
> 
> Sorry, I'm checking on it now

Thanks for the update (I'll go through the code once again today). May you just add some tests suggested in the code review comment?

hookSSi commented:
> Thanks for the update (I'll go through the code once again today). May you just add some tests suggested in the code review comment?

I added test that just checking value. Have a look and review me please.

Borda commented:
@hookSSi could you please mark addressed comments as resolved so we easily see what has been left? :rabbit:

hookSSi commented:
> @hookSSi could you please mark addressed comments as resolved so we easily see what has been left? 🐰

Oh I didn't know that I can mark these sorry about that.

Borda commented on tests/text/test_rouge.py:
```diff
@@ -164,3 +166,47 @@ def test_rouge_metric_wrong_key_value_error():
+    )
+    metrics_score = Scorer.compute()
+
+    threshold = 1e-08
+    np.isclose(metrics_score[rouge_level + "_" + metric], original_score, atol=threshold, equal_nan=True)
```
```suggestion
    np.isclose(metrics_score[rouge_level + "_" + metric], original_score, atol=1e-8, equal_nan=True)
```

Borda commented on tm_examples/rouge_score-own_normalizer_and_tokenizer.py:
```diff
@@ -0,0 +1,88 @@
+        """The `__call__` method must be defined for this class. To ensure the functionality, the `__call__` method
+        should obey the input/output arguments structure described below.
+
+        Args:
+            text:
+                Input text. `str`
```
```suggestion
        Args:
            text: Input text.
```

Borda commented on tm_examples/rouge_score-own_normalizer_and_tokenizer.py:
```diff
@@ -0,0 +1,88 @@
+class UserTokenizer:
+    """The `UserNormalizer` class is required to tokenize a non-alphabet language text input.
+
+    The user's defined tokenizer is expected to return `Sequence[str]` that are fed into the rouge score.
```
```suggestion
    The user's defined tokenizer is expected to return ``Sequence[str]`` that are fed into the rouge score.
```

Borda commented on tm_examples/rouge_score-own_normalizer_and_tokenizer.py:
```diff
@@ -0,0 +1,88 @@
+
+    The user's defined tokenizer is expected to return `Sequence[str]` that are fed into the rouge score.
+    """
+
+    def __init__(self) -> None:
+        self.pattern = r"\s+"
```
```suggestion
    pattern = r"\s+"
```
seems to be rather constant

Borda commented on tm_examples/rouge_score-own_normalizer_and_tokenizer.py:
```diff
@@ -0,0 +1,88 @@
+        should obey the input/output arguments structure described below.
+
+        Args:
+            text:
+                Input text. `str`
```
```suggestion
            text: Input text.
```
`str` comes from typing

Borda commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -101,17 +106,27 @@ def _normalize_and_tokenize_text(text: str, stemmer: Optional[Any] = None) -> Se
             An input sentence.
         stemmer:
             Porter stemmer instance to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This function must take a `str` and return a `str`.
+        tokenizer:
+            A user's own tokenizer function. If this is none, `spliting by spaces` is default
+            This function must take a `str` and return `Sequence[str]`
```
```suggestion
        normalizer:
            A user's own normalizer function.
            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
        tokenizer:
            A user's own tokenizer function. If this is ``None``, splitting by spaces is the default.
```
it is coming from typing

Borda commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -184,6 +201,13 @@ def _rouge_score_update(
             Allowed values are ``avg`` and ``best``.
         stemmer:
             Porter stemmer instance to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This function must take a `str` and return a `str`.
+        tokenizer:
+            A user's own tokenizer function. If this is none, `spliting by spaces` is default
+            This function must take a `str` and return `Sequence[str]`
```
```suggestion
        normalizer:
            A user's own normalizer function.
            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
        tokenizer:
            A user's own tokenizer function. If this is ``None``, splitting by spaces is the default.
```

Borda commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -306,6 +334,13 @@ def rouge_score(
             - ``best`` takes the best fmeasure score obtained between prediction and multiple corresponding references.
         use_stemmer:
             Use Porter stemmer to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This function must take a `str` and return a `str`.
+        tokenizer:
+            A user's own tokenizer function. If this is none, `spliting by spaces` is default
+            This function must take a `str` and return `Sequence[str]`
```
```suggestion
        normalizer:
            A user's own normalizer function.
            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
        tokenizer:
            A user's own tokenizer function. If this is ``None``, splitting by spaces is the default.
```

Borda commented on torchmetrics/text/rouge.py:
```diff
@@ -35,6 +35,13 @@ class ROUGEScore(Metric):
     Args:
         use_stemmer:
             Use Porter stemmer to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is none, `replacing any non-alpha-numeric characters with spaces` is default.
+            This function must take a `str` and return a `str`.
+        tokenizer:
+            A user's own tokenizer function. If this is none, `spliting by spaces` is default
+            This function must take a `str` and return `Sequence[str]`
```
```suggestion
        normalizer:
            A user's own normalizer function.
            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
        tokenizer:
            A user's own tokenizer function. If this is ``None``, splitting by spaces is the default.
```

stancld commented on tm_examples/rouge_score-own_normalizer_and_tokenizer.py:
```diff
@@ -0,0 +1,85 @@
+            text: Input text.
+
+        Return:
+            tokenized sentence
```
```suggestion
            Tokenized sentence
```

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -101,17 +106,25 @@ def _normalize_and_tokenize_text(text: str, stemmer: Optional[Any] = None) -> Se
             Porter stemmer instance to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
```
```suggestion
            If this is ``None``, replacing any non-alpha-numeric characters with space is the default.
```

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -184,6 +199,11 @@ def _rouge_score_update(
             Porter stemmer instance to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
```
```suggestion
            If this is ``None``, replacing any non-alpha-numeric characters with space is the default.
```

stancld commented on torchmetrics/functional/text/rouge.py:
```diff
@@ -306,6 +330,11 @@ def rouge_score(
             Use Porter stemmer to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
```
```suggestion
            If this is ``None``, replacing any non-alpha-numeric characters with space is the default.
```

stancld commented on torchmetrics/text/rouge.py:
```diff
@@ -35,6 +35,11 @@ class ROUGEScore(Metric):
             Use Porter stemmer to strip word suffixes to improve matching.
+        normalizer:
+            A user's own normalizer function.
+            If this is `None``, replacing any non-alpha-numeric characters with space` is the default.
```
```suggestion
            If this is ``None``, replacing any non-alpha-numeric characters with space is the default.
```

stancld reviewed (approved):
Nitpicks
